File: minimum-depth-of-binary-tree/minimum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDepth(self, root: TreeNode) -> int:
        if root is None:
            return 0
        min_depth = 1
        current_level, next_level = 1, 0
        queue = [root]
        while queue: # Run thru tree
            node = queue.pop(0)
            current_level -= 1
            if not node.left and not node.right: # Check if leaf node
                return min_depth
            if node.left: # Check left
                queue.append(node.left)
                next_level += 1
            if node.right: # Check right
                queue.append(node.right)
                next_level += 1
            if current_level == 0:
                min_depth += 1
                current_level, next_level = next_level, 0

File: minimum-depth-of-binary-tree/minimum-depth-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution 
{
    public int minDepth(TreeNode root) 
    {
        int depth = 1;
        if (root == null)
            return 0;
        
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        queue.add(root);
        int currentLevel = 1;
        int nextLevel = 0;
        while (currentLevel > 0)
        {
            TreeNode node = queue.remove();
            currentLevel--;
            
            // Found highest leaf node
            if (node.left == null && node.right == null)
                return depth;
            
            if (node.left != null)
            {
                queue.add(node.left);
                nextLevel++;
            }
            if (node.right != null)
            {
                queue.add(node.right);
                nextLevel++;
            }
            
            if (currentLevel == 0)
            {
                currentLevel = nextLevel;
                nextLevel = 0;
                depth++;
            }
        }
        
        return depth;
    }
}